src/collector.rs

```diff
@@ -1,63 +1,282 @@
 
-    for input in inputs {
-        match Url::parse(&input) {
-            Ok(url) => {
-                let path = String::from(url.path());
-                let res = reqwest::get(url).await?;
-                let content = res.text().await?;
+    // extract input contents
+    for input in inputs.iter().cloned() {
+        let mut sender = contents_tx.clone();
 
-                links.extend(extract_links(
-                    resolve_file_type_by_path(path),
-                    &content,
-                    base_url.clone(),
-                ));
-            }
-            Err(_) => {
-                // Assume we got a single file or a glob on our hands
-                for entry in glob(&input)? {
-                    match entry {
-                        Ok(path) => {
-                            let content = fs::read_to_string(&path)?;
-                            links.extend(extract_links(
-                                resolve_file_type_by_path(&path),
-                                &content,
-                                base_url.clone(),
-                            ));
-                        }
-                        Err(e) => println!("Error handling file pattern {}: {:?}", input, e),
-                    }
-                }
-            }
-        };
+        tokio::spawn(async move {
+            let contents = input.get_contents(None, skip_missing_inputs).await;
+            sender.send(contents).await
+        });
+    }
+
+    // receiver will get None once all tasks are done
+    drop(contents_tx);
+
+    // extract links from input contents
+    let mut extract_links_handles = vec![];
+
+    while let Some(result) = contents_rx.recv().await {
+        for input_content in result? {
+            let base_url = base_url.clone();
+            let handle =
+                tokio::task::spawn_blocking(move || extract_links(&input_content, base_url));
+            extract_links_handles.push(handle);
+        }
+    }
+
+    // Note: we could dispatch links to be checked as soon as we get them,
+    //       instead of building a HashSet with all links.
+    //       This optimization would speed up cases where there's
+    //       a lot of inputs and/or the inputs are large (e.g. big files).
+    let mut collected_links = HashSet::new();
+
+    for handle in extract_links_handles {
+        let links = handle.await?;
+        collected_links.extend(links);
+    }
+
+    Ok(collected_links)
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::test_utils::get_mock_server_with_content;
+    use std::fs::File;
+    use std::io::Write;
+    use std::str::FromStr;
+
+    const TEST_STRING: &str = "http://test-string.com";
+    const TEST_URL: &str = "https://test-url.org";
+    const TEST_FILE: &str = "https://test-file.io";
+    const TEST_GLOB_1: &str = "https://test-glob-1.io";
+    const TEST_GLOB_2_MAIL: &str = "test@glob-2.io";
+
+    #[tokio::test]
+    async fn test_collect_links() -> Result<()> {
```

Comment:
I think this test should be enough here, but I'm happy to add more (e.g. tests for all `Input` methods) if you think it makes sense.

Reply:
Looks perfectly fine to me. Thanks!